internal/pkg/agent/perms/windows.go

```diff
@@ -21,6 +21,15 @@ import (
 	"github.com/elastic/elastic-agent/pkg/utils"
 )
 
+func errHandler(err error) error {
```

Comment:
A clearer name would be better here. Being names `errHandler` doesn't really provide context to what it is doing, when reading the code where it is called from. Something like `filterSuccessErr` would be more appropriate.

```suggestion
func filterSuccessErr(err error) error {
```

Reply:
Should we put this on every windows sycall we make? What is special about `acl.Apply` here?

Reply:
I believe that this error code (ERROR_SUCCESS) handling should happen for all functions that call directly the Win32 low level API. Now acl.Apply is actually calling this API [here](https://github.com/hectane/go-acl/blob/master/api/secinfo.go#L71) but doesn't check for it. Thus I don't believe we should add this check in every windows syscall 🙂

Reply:
thank you @blakerouse I was clearly lacking inspiration when naming this function but I do like your proposal! I just committed it 🙂

Reply:
@blakerouse @cmacknz should I consider this conversation as resolved? 🙂

Reply:
Resolved from my side. I will let @cmacknz respond for his question on doing this with every Windows syscall we make.

Reply:
It looks to me like we are directly calling through to Windows DLLs in other places, like https://github.com/elastic/elastic-agent/blob/d6687876cb0ac8671719aecca7336cdeff671710/internal/pkg/agent/install/user_windows.go#L26-L32

Reply:
ok I believe then we should handle the errors, under this issue or a separate one?

Reply:
and after taping inside acl.Apply I believe we should re-implement that in our own. Specifically this is the body of acl.Apply

```
func Apply(name string, replace, inherit bool, entries ...api.ExplicitAccess) error {
	var oldAcl windows.Handle
	if !replace {
		var secDesc windows.Handle
		api.GetNamedSecurityInfo(
			name,
			api.SE_FILE_OBJECT,
			api.DACL_SECURITY_INFORMATION,
			nil,
			nil,
			&oldAcl,
			nil,
			&secDesc,
		)
		defer windows.LocalFree(secDesc)
	}
	var acl windows.Handle
	if err := api.SetEntriesInAcl(
		entries,
		oldAcl,
		&acl,
	); err != nil {
		return err
	}
	defer windows.LocalFree((windows.Handle)(unsafe.Pointer(acl)))
	var secInfo uint32
	if !inherit {
		secInfo = api.PROTECTED_DACL_SECURITY_INFORMATION
	} else {
		secInfo = api.UNPROTECTED_DACL_SECURITY_INFORMATION
	}
	return api.SetNamedSecurityInfo(
		name,
		api.SE_FILE_OBJECT,
		api.DACL_SECURITY_INFORMATION|secInfo,
		nil,
		nil,
		acl,
		0,
	)
}
```

SetEntriesInAcl [can return a ERROR_SUCCESS](https://learn.microsoft.com/en-us/windows/win32/api/aclapi/nf-aclapi-setentriesinaclw) which is not handled and instead it returns. Which means that we will miss calling `SetNamedSecurityInfoW` 🤦

Reply:
I could be wrong, but I think with that code snippet, `oldAcl` could have received a pointer to the DACL, in which case it would need to be freed when you exit.

Reply:
the more I read about it, the more I start to believe that we are mislead here gents. So our starting point is this error message `Error: failed to fix permissions: The operation completed successfully.` and I believe that exactly this one is derailing us. Specifically, this error comes from within `github.com/hectane/go-acl` from a function like this

```go
func SetNamedSecurityInfo(objectName string, objectType int32, secInfo uint32, owner, group *windows.SID, dacl, sacl windows.Handle) error {
	ret, _, err := procSetNamedSecurityInfoW.Call(
		uintptr(unsafe.Pointer(windows.StringToUTF16Ptr(objectName))),
		uintptr(objectType),
		uintptr(secInfo),
		uintptr(unsafe.Pointer(owner)),
		uintptr(unsafe.Pointer(group)),
		uintptr(dacl),
		uintptr(sacl),
	)
	if ret != 0 { 
		return err // <-- attention here actual ret is not returned
	}
	return nil
}
```

the godoc of the `.Call(...)` func states the following

```go
// Call executes procedure p with arguments a. It will panic, if more than 15 arguments
// are supplied.
//
// The returned error is always non-nil, constructed from the result of GetLastError (<-- pay attention here).
// Callers must inspect the primary return value to decide whether an error occurred
// (according to the semantics of the specific function being called) before consulting
// the error. The error will be guaranteed to contain windows.Errno.
```

Based on the above, I do believe that we indeed experience *an actual error* because `ret != 0`. In conjunction with the [official documentation of the function](https://learn.microsoft.com/en-us/windows/win32/api/aclapi/nf-aclapi-setnamedsecurityinfow) which doesn't mention that you should read the last error to see more details why `ret != 0`, I truly believe that the lastErr that `Call` returns and is propagated to us on `ret != 0` from the library regards another operation and not our error, thus it says `operation completed successfully.`. 

Based on all the above my proposal is actually do something similar we do for the other windows `Calls` in the code and propagate an error with the actual retval in it in the hope of being able understand better what is failing when this occurs

cc @cmacknz @blakerouse

Reply:
@leehinman for the code snippet comment; according to the [GetNamedSecurityInfoW documentation](https://learn.microsoft.com/en-us/windows/win32/api/aclapi/nf-aclapi-getnamedsecurityinfow) you shouldn't free `oldAcl`

```
[out, optional] ppDacl

A pointer to a variable that receives a pointer to the DACL in the returned security descriptor or NULL if the security descriptor has no DACL. The returned pointer is valid only if you set the DACL_SECURITY_INFORMATION flag. Also, this parameter can be NULL if you do not need the DACL.
```

where you should free `secDesc`

```
[out, optional] ppSecurityDescriptor

A pointer to a variable that receives a pointer to the security descriptor of the object. When you have finished using the pointer, free the returned buffer by calling the [LocalFree](https://learn.microsoft.com/en-us/windows/desktop/api/winbase/nf-winbase-localfree) function.
```

Reply:
gents after cross-validating my understanding with @blakerouse I think we both agree that there is an error happening and the first step to get more awareness of the root cause, is to propagate the "correct" error that captures the actual ret code that the Win32 API calls fail with. I have updated my PR to achieve that. So, I am resolving this one and let's kick off a new round of reviews please 🙂